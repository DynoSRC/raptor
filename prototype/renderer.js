const injecture = require('injecture');
const pug = require('pug');

// TODO: Raptor should handle the concept of dumb child views, not your
// renderer. Give people option of JSON vs. proto consumption.
const Renderer = class Renderer {
  // Your raptor.renderer interface must implement a render() method. It accepts
  // a View proto. Dumb child views are recursively rendered.
  // TODO: Make a generator version of this for the joke.
  async render(view, clientId, fragments=[]) {
    // If truthy, this is a boring View proto. We want the underlying view model
    // proto such as ExampleView or ExampleSetView.
    if (typeof view.hasView === 'function' && view.hasView()) {
      return await this.render(view.getView(), clientId, fragments);
    }

    const name = view.getContainer().getContainerName();
    // The top-level view model does not render itself, but is a container for
    // other views.
    // TODO: Enum or instanceof. No magic strings.
    if (name !== 'Model') {
      fragments.push(
          pug.renderFile(`prototype/templates/${name}.pug`, view.toObject()));
    }

    const childViews = view.getViewsList();
    if (childViews && childViews.length) {
      await Promise.all(childViews.map((view) => {
        return this.render(view, clientId, fragments);
      }));
    }

    return fragments.join('\n\n');
  }
};

injecture.registerClass(Renderer, {interfaces: ['raptor.renderer']});
exports.Renderer = Renderer;
