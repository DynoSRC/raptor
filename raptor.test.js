import React from 'react';
import Raptor from './raptor';
import renderer from 'react-test-renderer';

class TestApp extends React.Component {
  render() {
    return <React.Fragment>Sup I'm a JSX fragment.</React.Fragment>
  }
}

describe('raptor.views', () => {
  const Foo = {herp: 'derp'};
  const Bar = {derp: 'herp'};

  describe('addView', () => {
    it('Adds a single view to raptor.views.', () => {
      const raptor = new Raptor();
      raptor.addView('Foo', Foo);
      raptor.addView('Bar', Bar);
      expect(raptor.views).toEqual({Foo, Bar});
    });
  });

  describe('addViews', () => {
    it('Adds multiple views to raptor.views.', () => {
      const raptor = new Raptor();
      raptor.addViews({Foo, Bar});
      expect(raptor.views).toEqual({Foo, Bar});
    });
  });
});

describe('raptor.layouts', () => {
  const Foo = {herp: 'derp'};
  const Bar = {derp: 'herp'};

  describe('addLayout', () => {
    it('Adds a single layout to raptor.layouts.', () => {
      const raptor = new Raptor();
      raptor.addLayout('Foo', Foo);
      raptor.addLayout('Bar', Bar);
      expect(raptor.layouts).toEqual({Foo, Bar});
    });
  });

  describe('addLayouts', () => {
    it('Adds multiple layouts to raptor.layouts.', () => {
      const raptor = new Raptor();
      raptor.addLayouts({Foo, Bar});
      expect(raptor.layouts).toEqual({Foo, Bar});
    });
  });
});

describe('react-test-renderer', () => {
  it('Renders TestApp correctly.', () => {
    const component = renderer.create(<TestApp/>);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
