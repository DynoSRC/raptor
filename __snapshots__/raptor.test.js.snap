// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-test-renderer Renders TestApp correctly. 1`] = `"Sup I'm a JSX fragment."`;
